### ISSUE TITLE
Remove ES6 and templates in favor of strings and ES5.1

### ISSUE THREAD

**Issue by coolaj86** [AHAAAAAAA/PokemonGo-Map#1992]
There is almost nothing in the frontend code that requires an ES6 compiler (the only thing I see in map.js is string templates, which are very easy to do in ES5.1.

Can we just remove the ES6 so that the code will work in all modern browsers without the extra build step?

**Comment by megahambone** [comment]
+1 for this idea - but just a heads up, I've noticed there is something else in map.js that prevents it from running on Microsoft Edge properly (which normally handles ES6) - so there might be more to this than changing those template literals

**Comment by coolaj86** [comment]
What else is that?

**Comment by megahambone** [comment]
@coolaj86 - unfortunately, I'm not sure. All I know is that Edge worked up until map.js was fixed yesterday, now it doesn't. It says that the initmap is undefined, so there's some kind of code formatting or issue it doesn't like and it's apparently preventing the js from being parsed completely. What's more frustrating is that it intermittently loads the page in Edge, but it won't survive a F5 refresh ever.

**Comment by BillyWM** [comment]
@megahambone: That's due to loading order, not ES6. See PR https://github.com/AHAAAAAAA/PokemonGo-Map/pull/1869

The gmaps JS file is marked "async defer". Async, which takes precedence, says "load without blocking and run as soon as you can". The callback is initMap, which comes from map.js, loading separately. This is a race condition where if gmaps loads before map.js, it fails.

It happens intermittently because, as a race condition, either can win the race.

**Comment by BillyWM** [comment]
Removing ES6 wouldn't take away the extra build step. Grunt was already there doing some other tasks. Admittedly it isn't badly needed for those tasks, but it also isn't a huge deal to set up, and a lot of Javascript devs will have it already and already know how to use it.

This is really an ES5 vs ES6 discussion. More on that here: https://github.com/AHAAAAAAA/PokemonGo-Map/issues/1115

**Comment by BillyWM** [comment]
Also note that there's very little taking advantage of the ES6 compiler because it only recently got merged in. The PR to add it took a while before it was merged.

If it's kept around, we can put it to better use.

**Comment by megahambone** [comment]
@BillyWM - thanks for the info on the race condition. I'll probably resolve it in my fork (if it doesn't get resolved first)
Could you not implement some kind of shim or transpiler so that modern browsers that don't fully incorporate the ES6 featureset can still use the page, however? I'm all for code readability and neat syntax, but I really like compatibility with all the current gen browsers. That would satisfy the idea of this issue - compatibility while satisfying the nice syntax you're pitching. Seems win-win.

**Comment by coolaj86** [comment]
Grunt is just doing minification and sass

* minifying **1** javascript file **app.js** (that isn't even used, I don't think)
* sassifying **1** small sass file **main.sass**

**Comment by BillyWM** [comment]
@megahambone: That's what Babel is - a shim (babel-polyfill) and a transpiler (Babel itself), so that modern browsers that don't fully incorporate the ES6 featureset (All of them have gaps except Safari 10 on Mac) can still use the page.

@coolaj86 argues it's an unnecessary build step that raises the barrier to entry too high. I argue that it's not actually very difficult (less work than setting up Python properly, for sure!), and has meaningful benefits in the long run.

**Comment by coolaj86** [comment]
Let's say you were a teacher trying to show kids how to do fun cool computer stuff... don't you want them to simply be able to download and run the code?

Aside from that, it was only a few years ago that transpilers were the devil and people cared about this old fashioned ideals called "performance" and "stability" - the idea that you'd write code that would work everywhere now (and stillwork in 10 years), not write to some hypothetical virtual machine that hasn't actually been implemented to a spec that wasn't complete and has already had many things removed from it and many bugs found and huge performance hits causing code to run anywhere between 1/2 and 1/10 of the original speed.

But what do I know, I'm just a curmudgeonly old man...

**Comment by BillyWM** [comment]
No, none of that is interesting to me. The code should work, and competent developers should be able to contribute to it. I'm a curmudgeonly old man too and as such **I don't care about kids** :)

**Comment by coolaj86** [comment]
Well how about, performance, stability, and easy-of-use?

**Comment by BillyWM** [comment]
How does ES5 achieve that where ES6 doesn't? ES5 was born out of the aborted ES4 Harmony spec. It's hardly something to hold on a pedestal as well-crafted perfection, the highest form of the language, that we should never move beyond.

**Comment by BillyWM** [comment]
Again, this is just a programming language holy war. "I like it". "I don't". "Well I do". "Well I don't".

**Comment by coolaj86** [comment]
ES5 came from ES3

ES6 came from ES4 and CoffeeScript

**Comment by BillyWM** [comment]
No, ES5 came from nobody agreeing on what ES4 should be and putting out a more conservative update. Actionscript 3 took what would have been ES4 and ran with some bits of it.

ES6 is ES4 again now that we're ready for it.

There are a number of implementations with over 90% compatibility. In a few years ES6 will be the new ES5 and you'll be saying we should avoid ES7 and stick to tried and true ES6.

**Comment by BillyWM** [comment]
But fine, I'll get suckered in.

On ease of use:

Rest and spread are easier than `Array.prototype.slice.call(arguments)`

Sets and Maps are easier than doing it manually.

Arrow functions with lexical scoping are easier than `_self = this`

`const` and `let` just make better quality code.

`includes(x)` is easier than `indexOf(x) != -1`

Promises, generators, iterators blow callback-hell out of the water.

**Comment by coolaj86** [comment]
People!!! STOP TRYING TO MAKE JAVASCRIPT A CLASSICAL LANGUAGE!!!

Everyone seems to go back to this same idea that somehow you should be using classes in JavaScript. For those people I have two words: STOP IT!

JavaScript IS NOT OBJECT ORIENTED. It never has been. It never will be. Trying to do object-oriented stuff in an async language will NEVER WORK. It will always be more confusing. You can't call an `init` with async code inside that has to return an object before it's constructed and expect that to be fun.

Async + Stateful = BAD IDEA. Every. Time.

JavaScript is Procedural, Functional, and EVENT ORIENTED.

This is what JavaScript should look like:

```
var person = { username: 'user', password: 'secret' };
var Pokemon.login(person).then(function (session) {
  person.session = session;
  console.log(session);
  Pokemon.profile(person).then(function (profile) {
      person.profile = profile;
      console.log(profile);
  })
});
```

And then in the rare case that you need data privacy, you return a copy of the public data or you wrap some stuff in a closure.

The only thing JavaScript has to do with Java is the trademark. STOP THINKING ABOUT IT AS OBJECT-ORIENTED!!!

**Comment by bmds** [comment]
Just a quick note @BillyWM the Sass dependency in package.json is for [grunt-sass](https://github.com/sindresorhus/grunt-sass) which uses [libsass](http://sass-lang.com/libsass) (a c++ port of Sass) to compile. You won't be getting a Ruby dependency from that

**Comment by BillyWM** [comment]
>JavaScript IS NOT OBJECT ORIENTED.

Sure it is. It's just prototypal, not classical.

**Comment by BillyWM** [comment]
>This is what JavaScript should look like

That's just, like, you know, your opinion, man.

Promises are a step up from callback hell. I like them. But they still plunge you into *thenable hell*.  The easiest code to reason about is synchronous looking code (and the new syntax builds off Promises, btw). But that's just like my opinion man.

**Comment by BillyWM** [comment]
The Promise syntax may be compatible with ES5 but they didn't become part of the language until ES6. IE doesn't support them so we need a polyfill if we want to use them.
